app/services/ect_at_school_periods/switch_training.rb

```diff
@@ -0,0 +1,85 @@
+module ECTAtSchoolPeriods
+  class IncorrectTrainingProgrammeError < StandardError; end
+  class NoTrainingPeriodError < StandardError; end
+
+  class SwitchTraining
+    include TrainingPeriodSources
+
+    def self.to_school_led(...) = new(...).to_school_led
+    def self.to_provider_led(...) = new(...).to_provider_led
+
+    def initialize(ect_at_school_period, author:, lead_provider: nil)
+      raise ArgumentError, "an `ECTAtSchoolPeriod` must be provided" unless
+        ect_at_school_period.is_a?(ECTAtSchoolPeriod)
+
+      @ect_at_school_period = ect_at_school_period
+      @training_period = ect_at_school_period.current_or_next_training_period
+
+      raise NoTrainingPeriodError unless @training_period
+
+      @school = @ect_at_school_period.school
+      @lead_provider = lead_provider.presence || @training_period.lead_provider
+      @author = author
+    end
+
+    def to_school_led
+      raise IncorrectTrainingProgrammeError if
+        @ect_at_school_period.school_led_training_programme?
+
+      ActiveRecord::Base.transaction do
+        finish_or_destroy_existing_training_period!
+        create_school_led_training_period!
```

Comment:
Do we want to record an event when changing training programme?

Reply:
I agree yeah 👍🏽 

https://github.com/DFE-Digital/register-early-career-teachers-public/pull/1398#issuecomment-3318701449